network/transport/nodedid.go

```diff
@@ -0,0 +1,83 @@
+package transport
+
+import (
+	"github.com/nuts-foundation/go-did/did"
+	"github.com/nuts-foundation/nuts-node/crypto"
+	"github.com/nuts-foundation/nuts-node/vdr/doc"
+	"github.com/nuts-foundation/nuts-node/vdr/types"
+	"sync"
+	"time"
+)
+
+// NodeDIDResolver defines an interface for types that resolve the local node's DID, which is used to identify the node on the network.
+type NodeDIDResolver interface {
+	// Resolve tries to resolve the node DID. If it's absent, an empty DID is returned. In any other non-successful case an error is returned.
+	Resolve() (did.DID, error)
+}
+
+// FixedNodeDIDResolver is a NodeDIDResolver that returns a preset DID.
+type FixedNodeDIDResolver struct {
+	NodeDID did.DID
+}
+
+// Resolve returns the fixed-set node DID.
+func (f FixedNodeDIDResolver) Resolve() (did.DID, error) {
+	return f.NodeDID, nil
+}
+
+// NewAutoNodeDIDResolver creates a new node DID resolver that tried to look it up by matching DID documents with a NutsComm endpoint set and the private key of the local node.
+func NewAutoNodeDIDResolver(keyResolver crypto.KeyResolver, docFinder types.DocFinder) NodeDIDResolver {
+	return &AutoNodeDIDResolver{
+		keyResolver: keyResolver,
+		docFinder:   docFinder,
+		mux:         &sync.Mutex{},
+	}
+}
+
+type AutoNodeDIDResolver struct {
+	keyResolver crypto.KeyResolver
+	docFinder   types.DocFinder
+	mux         *sync.Mutex
+	resolvedDID did.DID
+}
+
+// Resolve returns the auto-resolved node DID, or an empty DID if none could be found.
+func (a *AutoNodeDIDResolver) Resolve() (did.DID, error) {
+	a.mux.Lock()
+	if !a.resolvedDID.Empty() {
+		result := a.resolvedDID
+		a.mux.Unlock()
+		return result, nil
+	}
+	defer a.mux.Unlock()
```

Comment:
Yes, and that's exactly why we should cache it. Otherwise the node DID might change during runtime. So it could present the node DID to one peer as `did:nuts:A` and to a second peer as `did:nuts:B`. Which would be very, very odd behavior imo.

Now it _shouldn't_ ever change, not even after a restart. But should it change (due to unforeseen flow/case), it's disallowed in strict mode (a.k.a. production usage).

Reply:
> Otherwise the node DID might change during runtime. So it could present the node DID to one peer as did:nuts:A and to a second peer as did:nuts:B. Which would be very, very odd behavior imo.

Why? The peer ID should be the 'main' identifier right? I agree that it shouldn't happen very often but it could happen and I think it should be supported.

Reply:
If we allow it to be changed at runtime (because a NutsComm endpoint is registered on an existing, or new DID document), which one should be matched as the one being the correct node DID? The behavior will be unclear at best (last updated DID document = node DID) or unpredictable at worst (random matching DID document).

I suppose it _could_ change in dev/workshop envs (because people will be fiddling around/losing keys/DIDs), but in that case I still think predictable, simple behavior is better than a maybe more intelligent but unpredictable algorithm that (maybe unexpectedly) changes the node DID halfway?

In production environments it must be configured statically anyways.